File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int total = 0;
        int leftMax = height[left];
        int rightMax = height[right];
        while (left < right) {
            if (leftMax > rightMax) {
                right--;
                rightMax = Math.max(rightMax, height[right]);
                total += rightMax - height[right];
            } else {
                left++;
                leftMax = Math.max(leftMax, height[left]);
                total += leftMax - height[left];
            }
        }
        return total;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // TC: O(n)
        // SC: O(1)
        int result = 0;
        int n = height.size();
        int left = 0;
        int right = n - 1;
        int leftMax = height[left];
        int rightMax = height[right];
        while (left < right) {
            if (leftMax > rightMax) {
                right--;
                rightMax = max(rightMax, height[right]);
                result += rightMax - height[right];
            } else {
                left++;
                leftMax = max(leftMax, height[left]);
                result += leftMax - height[left];
            }
        }
        return result;
    }
};